hw_diag/utilities/network_watchdog.py

```diff
@@ -0,0 +1,136 @@
+        self.logger.info(f"Network manager state: {self.network_manager.get_connect_state()}")
+        self.logger.info(f"Internet connectivity: {self.have_internet()}")
+
+        nm_connected = self.network_manager.is_connected()
+        return nm_connected
+
+    def restart_network_manager(self):
+        """Restart hostOS NetworkManager service"""
+        nm_restarted = self.network_manager_unit.wait_restart()
+        self.logger.info(f"Network manager restarted: {nm_restarted}")
+
+    def get_last_restart(self) -> datetime:
+        try:
+            last_restart = KeyStore(self.state_file_path).get(self.LAST_RESTART_KEY)
+            return datetime.strptime(last_restart, self.LAST_RESTART_DATE_FORMAT)
+        except Exception as e:
+            self.logger.info(f"Can not find the previous restart time: {e}")
+            return datetime.min
+
+    def save_last_restart(self) -> None:
+        store = KeyStore(self.state_file_path)
+        store.set(self.LAST_RESTART_KEY, datetime.now().strftime(self.LAST_RESTART_DATE_FORMAT))
+        self.logger.info("Saved the current time before restarting the hotpsot.")
+
+    def run_watchdog(self) -> None:
+        self.logger.info("Running the watchdog...")
+
+        if self.is_connected():
+            self.lost_count = 0
+            self.logger.info("Network is working.")
+        else:
+            self.lost_count += 1
+            self.logger.warning(
+                f"Network is not connected! Lost connectivity count={self.lost_count}")
+
+            if self.lost_count > self.NM_RESTART_THRESHOLD:
+                self.logger.warning(
+                    "Reached out to the lost connectivity count to restart the network manager.")
+                self.restart_network_manager()
+                self.logger.info("Restarted the network connection.")
+
+                if self.is_connected():
+                    self.lost_count = 0
+                    self.logger.info("Network is working after restarting the network connection.")
+                else:
+                    self.logger.warning("Network is still not working.")
+
+                    if self.lost_count > self.FULL_REBOOT_THRESHOLD:
+                        self.logger.warning(
+                            "Reached out to the lost connectivity count to reboot the hotspot.")
+                        if datetime.now() - timedelta(
+                                hours=self.REBOOT_LIMIT_HOURS) < self.get_last_restart():
+                            self.logger.info(
+                                "Hotspot has already been restarted within a day, skipping.")
+                        else:
+                            self.save_last_restart()
+                            self.logger.info("Rebooting the hotspot.")
+
+                            balena_supervisor = BalenaSupervisor.new_from_env()
+                            balena_supervisor.reboot(force=True)
```

Comment:
it might be useful to force if it is failing all the time. We have devices stuck in our fleets on updating. @KevinWassermann94 does a reboot recover such devices?

Reply:
@posterzh I still think if we keep a counter for balena_failed_reboot, we can assume that balena is stuck somewhere from which it cannot recover and we can force the reboot. A count of 3 which will be equivalent to 3 hours should be enough.

Reply:
Ok, that would work better.